lib/types.nix

```diff
@@ -1107,6 +1107,48 @@ rec {
     addCheck = elemType: check: elemType // { check = x: elemType.check x && check x; };
 
   };
+
+  /**
+    Merges two option types together.
+
+    :::{.note}
+    Uses the merge function of the first type, to merge it with the second type.
+
+    Usually types can only be merged if they are of the same type
+    :::
+
+    # Inputs
+
+    : `a` (option type): The first option type.
+    : `b` (option type): The second option type.
+
+    # Returns
+
+    - The merged option type.
+    - `null` if the types can't be merged.
```

Comment:
I'm a bit worried that this API isn't very future proof, because `null` doesn't allow for better error messages. How about we change it to return a string instead, for now just always "Cannot merge types"

Reply:
I made this an attribute set `{ error = ... ; }` for now.
I think this is the best compromise that is extensible in the future.

Error case alternatives:

- string: `Some Error`: Not extensible in the future. (Cannot hold other attributes)
- null: `null`: Leaks an implementation detail. Cannot hold any helpful information about the actual problem.
- Throw error: Downstream projects cannot handle error cases.